File: Java/62_Unique_Paths.java
class Solution {
    //global variable to store past results
    int mem[][];
    public int uniquePaths(int m, int n) {
        //create matrix
        mem = new int[m][n];
        //fill it with a filler variable
        for(int i= 0; i < m;i++) {
            Arrays.fill(mem[i], - 1);
        }
        //final spot in matrix we need to get to should be 1
        mem[m - 1][n - 1] = 1;
        //run recursive helper
        return memoize(0,0,m,n);
    }
    public int memoize(int i, int j, int m, int n) {
        //if we have an invalid index we exit
        if( i < 0 || j < 0 || i >= m || j >= n) {
            return 0;
        }
        //if we already computed these indices then we return past result
        if(mem[i][j] != -1) {
            return mem[i][j];
        }
        //otherwise we calculate both going right and going down
        int right = memoize(i, j + 1,m,n);
        int down = memoize(i + 1,j ,m,n);
        //store it in our matrix
        mem[i][j] = right + down;
        //and return it :)
        return mem[i][j];
    }
    
    
}

File: Python/62_Unique_Paths.py
class Solution(object):
    def uniquePaths(self, m, n):
        """
        :type m: int
        :type n: int
        :rtype: int
        """
        lst = [[-1 for _ in range(n)] for _ in range(m)]
        lst[0][0] = 1
        def uniqueRec(i, j):
            if i < 0 or i >= m:
                return 0
            if j < 0 or j >= n:
                return 0
            if lst[i][j] != -1:
                return lst[i][j]
            lst[i][j] = uniqueRec(i - 1, j) + uniqueRec(i, j - 1)
            return lst[i][j]
        return uniqueRec(m - 1, n - 1)
            

        
